### PR TITLE
 fate test for dnn_module v3

### PR DESCRIPTION
**Pull request by Avery-xu**

### CODE REVIEW

File: libavfilter/dnn/dnn_backend_native_layer_conv2d.c [hunk 1]
```diff
@@ -152,35 +211,35 @@ static void * dnn_execute_layer_conv2d_thread(void *threadarg)
...
-                                int x_pos = CLAMP_TO_EDGE(x + (kernel_x - radius) * conv_params->dilation, width);
-                                input_pel = input[y_pos * src_linesize + x_pos * conv_params->input_num + ch];
-                            } else {
-                                int y_pos = y + (kernel_y - radius) * conv_params->dilation;
-                                int x_pos = x + (kernel_x - radius) * conv_params->dilation;
-                                input_pel = (x_pos < 0 || x_pos >= width || y_pos < 0 || y_pos >= height) ? 0.0 :
-                                                   input[y_pos * src_linesize + x_pos * conv_params->input_num + ch];
-                            }
-
+                    output[n_filter] += conv_params->biases[n_filter];
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d.c]: why we don't put this into _c function.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d.c [hunk 2]
```diff
@@ -223,6 +282,7 @@ int dnn_execute_layer_conv2d(DnnOperand *operands, const int32_t *input_operand_
     thread_common_param.output_operand_index = output_operand_index;
     thread_common_param.parameters = parameters;
     thread_common_param.ctx = ctx;
+
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d.c:202]: don't add extra empty line

**Review by guoyejun — COMMENTED** [review]
-cpuflags flags (global)
Allows setting and clearing cpu flags. This option is intended for testing. Do not use it unless you know what you’re doing.

don't add -cpuflags 0x100 in the commit log since it is for testing.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d.c [hunk 3]
```diff
@@ -233,7 +234,10 @@ static void * dnn_execute_layer_conv2d_thread(void *threadarg)
     exe_param.filter_size = filter_size;
     exe_param.filter_linesize = filter_linesize;
 
-     if ((thread_common_param->channel_step >= 4) && (conv_params->input_num >= 4)) {
+    if ((thread_common_param->channel_step >= 8) && (conv_params->input_num >= 8)) {
+        ff_dnn_execute_layer_conv2d_avx2(&exe_param);
+    }
+    else if ((thread_common_param->channel_step >= 4) && (conv_params->input_num >= 4)) {
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d.c]: if () {

} else if () {

} else {

}

File: libavfilter/dnn/dnn_backend_native_layer_conv2d.c [hunk 4]
```diff
@@ -296,6 +300,8 @@ int dnn_execute_layer_conv2d(DnnOperand *operands, const int32_t *input_operand_
         int cpu_flags = av_get_cpu_flags();
         if (EXTERNAL_SSE4(cpu_flags))
             thread_common_param.channel_step = 4;
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d.c]: what does channel mean for cpu flag?

a better naming might be asm_type or something else, and do the judgement here, like below:

if ((EXTERNAL_SSE4(cpu_flags) && (conv_params->input_num >= 4))
   thread_common_param.asm_type = SSE.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    .count_end:
+%endmacro
+
+; void ff_dnn_execute_layer_conv2d_sse4(const execute_param *exe_param);
+
+%if ARCH_X86_64
+INIT_XMM sse4
+cglobal dnn_execute_layer_conv2d, 8, 15, 3, exe_param,\
+    x, y, n_filter, cha, kernel_x, kernel_y, x_pos, y_pos, kernel_pos,\
+    input, output, kernel, tmp1, tmp2
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: %macro cglobal 1-2+ "" ; name, [PROLOGUE args]

why we need to add x, y, n_filter, ... which are not function args, they are also not used in the below code.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+%define src_linesize [exe_paramq + 11 * 4]
+%define filter_size [exe_paramq + 12 * 4]
+%define filter_linesize [exe_paramq + 13 * 4]
+%define SAME_CLAMP_TO_EDGE 2
+
+    mov inputq, [exe_paramq + 14 * 4]
+    mov outputq, [exe_paramq + 14 * 4 + 8]
+    mov kernelq, [exe_paramq + 14 * 4 + 2 * 8]
+
+    mov yd, thread_start
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: what do the suffix 'q' and 'd' mean?

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+.loop_y:
+    mov xd, pad_size
+    .loop_x:
+        xor n_filterd, n_filterd
+        xor kernel_posq, kernel_posq
+        .loop_filter:
+            xorps m2, m2
+            xor kernel_yd, kernel_yd
+
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: why empty line here and there? 

better to add comments for a different logic block.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            mov tmp1d, kernel_yd
+            sub tmp1d, radius
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: possible to have a better naming for m0, m1, m2, tmp1d and tmp2d ...?

File: libavfilter/dnn/dnn_backend_native_layer_conv2d.c (hunk 1 above)
Comment (Avery-xu) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d.c]: This should be put into _c function and can be used as initial value of output.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm (hunk 5 above)
Comment (Avery-xu) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: x, y, n_filter were names of registers, which can be seen as variables.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm (hunk 6 above)
Comment (Avery-xu) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: d, q were data length of the registers. d means the lower 32-bit of a 64-bit register, and q means the whole register.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm (hunk 8 above)
Comment (Avery-xu) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: m0, m1,..., were vector registers. In XMM case, they are 128-bit width. Usually we don't rename them. As for tmp1 and tmp2, I will get a new name for them.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm (hunk 5 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: but there variables are not used in the below code.

File: libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm (hunk 8 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_native_layer_conv2d_x86.asm]: thanks, and also add comments if necessary to help reviewers to understand easily.

**Pull request closed without merging**